src/pip/_internal/req/req_install.py

```diff
@@ -169,11 +170,15 @@ def from_req(cls, req, comes_from=None, isolated=False, wheel_cache=None):
             req = Requirement(req)
         except InvalidRequirement:
             raise InstallationError("Invalid requirement: '%s'" % req)
-        if req.url:
+
+        if req.url and comes_from.link.netloc == PyPI.file_storage_domain:
```

Comment:
This should also check for TestPyPI though.

Suggestion (you can use a different way if you like):

```py3
domains_not_allowed = [PyPI..., TestPyPI...]
comes_from_domain = comes_from...
if req.url and comes_from_domain in domains_not_allowed:
    ...
```